keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+container fileystem.  Kubelet handles generating the key, requesting a
+certificate from the named signer, and renewing the certificate before it
+expires.
+
+Kubelet supports only a few key generation strategies, each named by a string
+that encapsulates both the key type and any parameters necessary (for example,
+RSA modulus size).  The intention is to offer a a tasting menu of reasonable key
+choices, rather than offering a flexibility to a wide variety of parameters.
+Key types supported by Kubelet are "RSA2048", "RSA3072", "RSA4096", "ECDSAP256",
+and "ECDSAP384".  If no key type is specified, Kubelet defaults to "ECDSAP256".
+
+<<[UNRESOLVED]>>
+Should RSA2048 be offered?  I'm inclined to say no.  Are there any other key types that should be offered?
+<<[/UNRESOLVED]>>
+
+Once the key is generated, Kubelet creates a CertificateSigningRequest that
+contains an X.509 CertificateRequest with no data in the Name or
+SubjectAlternateName fields, but instead embeds an X.509 PodIdentity extension
+with the details of the pod that mounted the volume.  The signer is intended to treat this CertificateRequest merely as an indication that it should issue a certificate for the named pod, rather than trying to issue a certificate that follows the structure of the CertificateRequest.
+
+Kubelet then starts a single-object informer on the CertificateSigningRequest object it just created, and waits until the CSR is Approved and Issued.  It then takes the private key it generated and the issued certificate chain, and writes them to the pod filesystem in a single PEM file.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the CSR.
+
+Kubelet strips any inter-block comments and intra-block headers that were
+present in the issued certificate chain from the signer.
+
+Kubelet always sets a 24-hour TTL on its certificate requests, although of course the signer is free to ignore the requested TTL and return a certificate with any TTL.
+
+During the initial volume mount, before the pod has started, any errors returned during the certificate issuance process will block pod startup.
+
+Kubelet will begin trying to rotate the certificate once less than 50% of its TTL is remaining, or after 24 hours, whichever is sooner.  Errors in the certificate issuance process are treated as non-blocking while the original certificate is still valid.  Once the original certificate has expired, errors are treated as blocking, because it's better for the pod to noisily go unhealthy than to continue to run with an expired certificate.
+
+<<[UNRESOLVED]>>
+The rotation/renewal behavior is minimally tunable.  Is greater
+tunability needed?  For example, is it necessary to let the pod author set the
+TTL Kubelet passes in the CSR object, or to let the pod author control the
+rotation safety threshold?
+<<[/UNRESOLVED]>>
+
```

Comment:
How does the kubelet replace the PEM-encoded file / do we specify any level of atomicity?

Reply:
All projected volume content is written using an atomic symlink directory swap using the `atomicwriter` library.

Reply:
Ok, cool.